### PR TITLE
Sigoptlite fix best assignments

### PR DESCRIPTION
**Pull request by erichanslee**

### CODE REVIEW

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 1]
```diff
@@ -127,108 +143,150 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
 
-    for best_observation in best_observations_from_logger:
-      for i, metric in enumerate(best_observation["values"]):
+    for best_assignment in best_assignments_list:
+      for i, value in enumerate(best_assignment.values):
         if objectives[i] == "maximize":
-          assert metric["value"] >= thresholds[i]
+          assert value.value >= thresholds[i]
         if objectives[i] == "minimize":
-          assert metric["value"] <= thresholds[i]
```
Comment (malkomes) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:189]: you can remove `thresholds` and `objectives` and just loop over the metrics here:
```
for best_assignment in best_assignments_list:
  for i, metric in enumerate(experiment_meta["metrics"]):
    assert metric["name"] == best_assignment.values[i].name
    if metric["objective"] ....
```
also it makes explicit that the order of metric values should be the same as the experiment.metrics

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 2]
```diff
@@ -5,51 +5,52 @@
 
 import numpy
 import pytest
+from sigopt import Connection
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:8]: you have to change all the

```python
with pytest.raises(ValueError) as ...
```
to 
```python
from sigopt.exception import SigOptException

with pytest.raises(SigOptException) as ...

```

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 3]
```diff
@@ -127,108 +143,150 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+    for _ in range(num_observations):
+      suggestion = self.conn.experiments(e.id).suggestions().create()
+      if suggestion.task.cost < 1:
+        values = [dict(name="y1", value=wrong_optimal_value)]
+      else:
+        values = [dict(name="y1", value=numpy.random.rand())]
+      self.conn.experiments(e.id).observations().create(
+        suggestion=suggestion.id,
+        values=values,
+        task=suggestion.task,
```
Comment (malkomes) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:254]: i think you can just insert the observation with high cost manually
that will remove the randomness

**Review by malkomes — APPROVED** [review]
couple of minors

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 4]
```diff
@@ -64,59 +65,74 @@ def create_observations_with_values_in_simplex(self, experiment, on_boundary=Fal
...
-    self.assert_observation_lists_are_equal(best_observations_from_logger, expected_pareto_optimal_observations)
+    experiment_meta["type"] = "random"
+    e = self.conn.experiments().create(**experiment_meta)
+
+    num_optimal = 12
+    pareto_optimal_values = self.create_values_in_simplex(on_boundary=True, num=num_optimal)
+    non_optimal_values = self.create_values_in_simplex(on_boundary=False)
+    observation_values_list = pareto_optimal_values + non_optimal_values
+    assignments_list = [
+      self.conn.experiments(e.id).suggestions().create().assignments for _ in range(len(observation_values_list))
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:119]: This is not the same thing as before, since all assignments would be exactly the same.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 5]
```diff
@@ -127,108 +143,150 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
-    best_assignments_logger = BestAssignmentsLogger(experiment)
-    best_observations_from_logger = best_assignments_logger.fetch(observations)
-    assert len(best_observations_from_logger) == 1
-    assert best_observations_from_logger[0]["task"]["cost"] == 1
-    assert best_observations_from_logger[0]["task"]["name"] == "expensive"
+    best_assignments = self.conn.experiments(e.id).best_assignments().fetch()
+    best_assignments_list = list(best_assignments.iterate_pages())
+    assert len(best_assignments_list) == 1
+
+    self.conn.experiments(e.id).observations().fetch()
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:261]: What's the point of this line?

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 6]
```diff
@@ -127,108 +143,150 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+    assert best_assignments_list[0].values[0].value != 200
+    assert best_assignments_list[0].values[0].value <= 1
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:263]: line 263 implies line 263 is true?

Need to verify that task is the full cost one.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py (hunk 3 above)
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:254]: we don't support manually inserting tasks

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 7]
```diff
@@ -5,51 +5,52 @@
...
-# TODO: tests for multisolution
 class TestBestAssignmentsLogger(UnitTestsBase):
+  conn = Connection(driver=LocalDriver)
+
+  @staticmethod
+  def best_assignments_to_value_list(best_assignments):
+    best_values_seen = []
+    for best_assignment in best_assignments:
+      best_values_seen.append([dict(name=value.name, value=value.value) for value in best_assignment.values])
+    return best_values_seen
```
[the remark this answers is not in the document]
Comment (harvey-intel) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:23]: We should also be matching the assignment dict as well whenever we could?

**Review by harvey-intel — CHANGES_REQUESTED** [review]
There are still some parts missing here.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py (hunk 2 above)
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:8]: I don't do any of the value error stuff here unless I'm missing something

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py (hunk 6 above)
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:255]: the annoying bit is that by design, best assignments does not tell you the task information (same for sigopt server), so not sure how I'll check... let me think about it

Comment (harvey-intel) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:255]: Ah I see. so we always assume that the best assignments is at full_cost when returned to the user.

Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:255]: yea I thought I was losing my mind yesterday for a bit, but I guess that's how it is.

Working on verifying assignments too

Comment (harvey-intel) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:255]: One idea is to just `to_json()` both objects and compare the `assignments` and `metrics` fields.

Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:255]: I think I'm going to keep the test as is, it does its job and verifies that only the full cost tasks are being output by best assignments (even if the explicit task cost isn't being attached)

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 8]
```diff
@@ -64,59 +83,76 @@ def create_observations_with_values_in_simplex(self, experiment, on_boundary=Fal
   def test_failed_observations(self, feature):
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:83]: probably don't need to test this for 5 features.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 9]
```diff
@@ -127,108 +163,146 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
-    for observation in expected_pareto_optimal_observations:
-      if observation["values"][0]["value"] >= threshold_y1 and observation["values"][1]["value"] <= threshold_y2:
-        assert observation in best_observations_from_logger
+    for best_assignment in best_assignments_list:
+      best_values = [dict(name=value.name, value=value.value) for value in best_assignment.values]
+      if best_values[0]["value"] >= threshold_y1 and best_values[1]["value"] <= threshold_y2:
+        assert best_values in pareto_optimal_values
       else:
-        assert observation not in best_observations_from_logger
+        assert best_values not in pareto_optimal_values
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:198]: we wouldn't have this `else` in best_assignments, since all values need satisfy the thresholds.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 10]
```diff
@@ -127,108 +163,146 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
-    best_observations_from_logger = best_assignments_logger.fetch(observations)
-    self.assert_observations_are_sorted(best_observations_from_logger, experiment)
-
-    expected_pareto_optimal_observations = [o.get_client_observation(experiment) for o in pareto_optimal_observations]
-    for observation in expected_pareto_optimal_observations:
-      if observation["values"][0]["value"] >= threshold_y1 and observation["values"][1]["value"] <= threshold_y2:
-        assert observation in best_observations_from_logger
+    for best_assignment in best_assignments_list:
+      best_values = [dict(name=value.name, value=value.value) for value in best_assignment.values]
+      if best_values[0]["value"] >= threshold_y1 and best_values[1]["value"] <= threshold_y2:
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:195]: this would need to be an assert.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 11]
```diff
@@ -127,108 +163,146 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+      suggestion = self.conn.experiments(e.id).suggestions().create()
+      self.conn.experiments(e.id).observations().create(
+        suggestion=suggestion.id,
+        values=[dict(name="y1", value=numpy.random.rand()), dict(name="y2", value=numpy.random.rand())],
+      )
+
+    best_assignments = self.conn.experiments(e.id).best_assignments().fetch()
+    best_assignments_list = list(best_assignments.iterate_pages())
+    self.assert_best_assignments_are_sorted(e, best_assignments_list)
+    assert len(best_assignments_list) > 1
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:246]: this is probably fine, the chance of this being 0 or 1 is like (0.75 * 0.25)^n  or (n-1).

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 12]
```diff
@@ -127,108 +163,146 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
-        assert observation in best_observations_from_logger
+    for best_assignment in best_assignments_list:
+      best_values = [dict(name=value.name, value=value.value) for value in best_assignment.values]
+      if best_values[0]["value"] >= threshold_y1 and best_values[1]["value"] <= threshold_y2:
+        assert best_values in pareto_optimal_values
       else:
-        assert observation not in best_observations_from_logger
+        assert best_values not in pareto_optimal_values
 
   def test_metric_constraints(self):
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:200]: For metric constraints, we should also test the trivial case of non of the observations satisfying the constraint, where we return an empty list.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 13]
```diff
@@ -127,108 +163,146 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+        optimized_values_within_threshold.append(observation.values[1].value)
 
-    valid_observations = best_assignments_logger.filter_valid_full_cost_observations(observations)
-    expected_valid_observations = [o.get_client_observation(experiment) for o in valid_observations]
-    for observation in expected_valid_observations:
-      assert best_observation["values"][1]["value"] <= observation["values"][1]["value"]
+    for optimized_value in optimized_values_within_threshold:
+      assert best_observation.values[1].value <= optimized_value
 
   def test_multimetric_search(self):
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:230]: Also test for no observations satisfying the thresholds.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 14]
```diff
@@ -127,108 +163,146 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+    for observation in observations_list:
+      if observation.values[0].value >= threshold:
+        optimized_values_within_threshold.append(observation.values[1].value)
 
-    valid_observations = best_assignments_logger.filter_valid_full_cost_observations(observations)
-    expected_valid_observations = [o.get_client_observation(experiment) for o in valid_observations]
-    for observation in expected_valid_observations:
-      assert best_observation["values"][1]["value"] <= observation["values"][1]["value"]
+    for optimized_value in optimized_values_within_threshold:
+      assert best_observation.values[1].value <= optimized_value
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:228]: can combine these two to just
```
if observation.values[0].value >= threshold:
      observation.values[1].value >= best_observation.values[1].value
```

**Review by harvey-intel — CHANGES_REQUESTED** [review]
Mostly fine with what you have here. There are a few more empty list edge cases that we need to test.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 15]
```diff
@@ -64,59 +83,76 @@ def create_observations_with_values_in_simplex(self, experiment, on_boundary=Fal
...
-    best_observations_from_logger = best_assignments_logger.fetch(observations)
-    self.assert_observations_are_sorted(best_observations_from_logger, experiment)
-    expected_pareto_optimal_observations = [o.get_client_observation(experiment) for o in pareto_optimal_observations]
-    self.assert_observation_lists_are_equal(best_observations_from_logger, expected_pareto_optimal_observations)
+    experiment_meta["type"] = "random"
+    e = self.conn.experiments().create(**experiment_meta)
+
+    num_optimal = 12
+    pareto_optimal_assignments = self.generate_random_assignments(experiment_meta, num=num_optimal)
+    non_optimal_assignments = self.generate_random_assignments(experiment_meta, num=num_optimal)
```
[the remark this answers is not in the document]
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:135]: removed num_optimal from non_optimal_assignments

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 16]
```diff
@@ -64,59 +83,76 @@ def create_observations_with_values_in_simplex(self, experiment, on_boundary=Fal
...
+    assignments_list = pareto_optimal_assignments + non_optimal_assignments
+
+    pareto_optimal_values = self.create_values_in_simplex(on_boundary=True, num=num_optimal)
+    non_optimal_values = self.create_values_in_simplex(on_boundary=False)
```
[the remark this answers is not in the document]
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:139]: removed num_optimal from non_optimal_assignments

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py (hunk 9 above)
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:198]: removed else statement, changed if statement to assertion that values satisfy thresholds

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py (hunk 11 above)
Comment (erichanslee) [reply sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:246]: yea that's why I jacked up the number of observations

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 17]
```diff
@@ -128,107 +160,155 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+      ],
+      [
+        dict(name="y1", objective="maximize", strategy="constraint", threshold=1),
+        dict(name="y2", objective="minimize", strategy="constraint", threshold=1),
+      ],
+    ],
+  )
+  def test_multimetric_all_points_unsatisfactory_empty_list(self, metrics):
+    experiment_meta = dict(
+      parameters=DEFAULT_PARAMETERS,
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:275]: can use `DEFAULT_SIMPLE_PARAMETERS` to make this test even lighter.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 18]
```diff
@@ -128,107 +160,155 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
-    best_observations_from_logger = best_assignments_logger.fetch(observations)
-    self.assert_observations_are_sorted(best_observations_from_logger, experiment)
-
-    expected_pareto_optimal_observations = [o.get_client_observation(experiment) for o in pareto_optimal_observations]
-    for observation in expected_pareto_optimal_observations:
-      if observation["values"][0]["value"] >= threshold_y1 and observation["values"][1]["value"] <= threshold_y2:
-        assert observation in best_observations_from_logger
-      else:
-        assert observation not in best_observations_from_logger
+    for best_assignment in best_assignments_list:
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:182]: can also `assert len(best_assignments_list) < num_optimal`, since we are thresholding a portion of the pareto frontier.

File: sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py [hunk 19]
```diff
@@ -128,107 +160,155 @@ def test_multimetric_thresholds(self, threshold_y1, threshold_y2):
...
+      [dict(name="y1", objective="maximize", threshold=1), dict(name="y2", objective="maximize", threshold=1)],
+      [
+        dict(name="y1", objective="maximize", strategy="constraint", threshold=1),
+        dict(name="y2", objective="minimize", strategy="optimize"),
+      ],
+      [
+        dict(name="y1", objective="maximize", strategy="constraint", threshold=1),
+        dict(name="y2", objective="minimize", strategy="constraint", threshold=1),
+      ],
+    ],
```
Comment (harvey-intel) [inline sigoptlite/sigoptlitetest/endpoints/test_best_assignments.py:271]: These test check are a bit cheeky; they depend on that you are generating y1, y2 in [0, 1]. 

I think a better way of setting up this test is to create the experiments as defined in the feature constants and then manually create a few observations with values that violates 1 or all of the thresholds (which you can access via either the meta or the experiment object).

**Review by harvey-intel — APPROVED** [review]
Mostly LGTM. A few more suggestions for improvement.
